File: week2_priority_queues_and_disjoint_sets/2_job_queue/job_queue.py
# python3

from collections import namedtuple
import heapq as pq

AssignedJob = namedtuple("AssignedJob", ["worker", "started_at"])
class Thread:
    def __init__(self, ID):
        self.ID = ID
        self.next_free_time = 0

    # Refer to stack overflow thread at:
    # https://stackoverflow.com/questions/8875706/heapq-with-custom-compare-predicate
    def __lt__(self, other):
        # If two threads becomes free concurrently, then the thread with smaller ID
        # takes precedence over the one with larger ID

        # determine who is of higher priority, which is different from C++
        if self.next_free_time == other.next_free_time:
            return self.ID < other.ID   # lager ID lower priority

        return self.next_free_time < other.next_free_time  # large next_free_time lower priority

class Job_Queue:
    def __init__(self):
        self.result = []
        self.n_jobs = 0
        self.n_workers = 0
        self.jobs = []

    def read_data(self):
        self.n_workers, self.n_jobs = map(int, input().split())
        self.jobs = list(map(int, input().split()))
        assert len(self.jobs) == self.n_jobs

    def assign_jobs_fast(self):
        # namedtuple is also immutable just like tuple
        # for i in range(self.n_jobs):
        #     self.result.append(AssignedJob(0, 0))

        self.pq_threads = []
        for i in range(self.n_workers):
            pq.heappush(self.pq_threads, Thread(i))

        for i in range(self.n_jobs):
            thread_free = pq.heappop(self.pq_threads)  # return and remove ele with highest priority
            self.result.append(AssignedJob(thread_free.ID, thread_free.next_free_time))
            thread_free.next_free_time += self.jobs[i]
            pq.heappush(self.pq_threads, thread_free)

    def write_response(self):
        for job in self.result:
            print(job.worker, job.started_at)

    def Solve(self):
        self.read_data()
        self.assign_jobs_fast()
        self.write_response()


# def assign_jobs(n_workers, jobs):
#     # TODO: replace this code with a faster algorithm.
#     result = []
#     next_free_time = [0] * n_workers
#     for job in jobs:
#         next_worker = min(range(n_workers), key=lambda w: next_free_time[w])
#         result.append(AssignedJob(next_worker, next_free_time[next_worker]))
#         next_free_time[next_worker] += job

#     return result


def main():
    job_queue = Job_Queue()
    job_queue.Solve()
    # n_workers, n_jobs = map(int, input().split())
    # jobs = list(map(int, input().split()))
    # assert len(jobs) == n_jobs

    # assigned_jobs = assign_jobs(n_workers, jobs)

    # for job in assigned_jobs:
    #     print(job.worker, job.started_at)


if __name__ == "__main__":
    main()

File: week2_priority_queues_and_disjoint_sets/2_job_queue/job_queue.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>   // std:priority_queue

using namespace std;
class Thread 
{ 
  public:
    // One thread has two data members: ID and the time when it becomes free again.
    int ID;
    long long next_free_time;

    // Constructor
    Thread(int _ID)
      :ID(_ID)
    {
      next_free_time = 0;
    }
};


class Compare_Thread
{
  public:
    // determine who is of lower priority
    bool operator () (Thread & thread1, Thread &thread2)
    {
      if(thread1.next_free_time == thread2.next_free_time)
      {
	// For the two threads who become free simultaneously, 
	// the tread with smaller ID is given precedence over the one with larger ID.
	return thread1.ID >  thread2.ID;
      }

      return thread1.next_free_time > thread2.next_free_time;
    }
};

class JobQueue {
 private:
  int num_workers_;
  vector<int> jobs_;

  vector<int> assigned_workers_;
  vector<long long> start_times_;

  void WriteResponse() const {
    for (int i = 0; i < jobs_.size(); ++i) {
      cout << assigned_workers_[i] << " " << start_times_[i] << "\n";
    }
  }

  void ReadData() {
    int m;
    cin >> num_workers_ >> m;
    jobs_.resize(m);
    for(int i = 0; i < m; ++i)
      cin >> jobs_[i];
  }

  void AssignJobs() {
    // TODO: replace this code with a faster algorithm.
    assigned_workers_.resize(jobs_.size());
    start_times_.resize(jobs_.size());
    vector<long long> next_free_time(num_workers_, 0);
    for (int i = 0; i < jobs_.size(); ++i) {
      int duration = jobs_[i];
      int next_worker = 0;
      for (int j = 0; j < num_workers_; ++j) {
        if (next_free_time[j] < next_free_time[next_worker])
          next_worker = j;
      }
      assigned_workers_[i] = next_worker;
      start_times_[i] = next_free_time[next_worker];
      next_free_time[next_worker] += duration;
    }
  }

  void assign_jobs_fast()
  {
    /*
       In contrast with the naive implementation, we replace vector of threads with priority_queue of threads to perform the search for suitable worker in logarithmic time.
       Hence, the total running time falls/drops from O(n^2) to  O(nlogn)
    */ 
    // Info to be stored
    assigned_workers_.resize(jobs_.size());
    start_times_.resize(jobs_.size());

    priority_queue<Thread, vector<Thread>, Compare_Thread> pq_threads;
    for(int i(0); i < num_workers_; i++)
    {
      pq_threads.push(Thread(i));
    }

    for(int i(0); i < jobs_.size(); i ++)
    {
      int duration = jobs_[i];
      Thread thread_free = pq_threads.top();
      pq_threads.pop();
      assigned_workers_[i] = thread_free.ID;
      start_times_[i] = thread_free.next_free_time;
      thread_free.next_free_time += duration;
      pq_threads.push(thread_free);
    }
  }

 public:
  void Solve() {
    ReadData();
    /* AssignJobs(); */
    assign_jobs_fast();
    WriteResponse();
  }
};

int main() {
  std::ios_base::sync_with_stdio(false);
  JobQueue job_queue;
  job_queue.Solve();
  return 0;
}
